audio/drivers/coreaudio.cpp

```diff
@@ -0,0 +1,466 @@
+      
+      desc.componentType = kAudioUnitType_Output;
+      desc.componentSubType = kAudioUnitSubType_HALOutput;
+      desc.componentManufacturer = kAudioUnitManufacturer_Apple;
+      desc.componentFlags = 0;
+      desc.componentFlagsMask = 0;
+
+      comp = AudioComponentFindNext(NULL, &desc);
+      if (!comp) {
+            qDebug("AUHAL component not found.");
+            *audioUnit = NULL;
+            *audioDevice = kAudioDeviceUnknown;
+            return false;
+            }
+
+      OSStatus status = AudioComponentInstanceNew(comp, audioUnit);
+      if (status) {
+            qDebug("Failed to open AUHAL component.");
+            *audioUnit = NULL;
+            *audioDevice = kAudioDeviceUnknown;
+            return false;
+            }
+
+      status = AudioUnitSetProperty(*audioUnit, kAudioOutputUnitProperty_CurrentDevice, kAudioUnitScope_Global, 0, audioDevice, sizeof(AudioDeviceID));
+      if (status) {
+            qDebug("Failed to set current device.");
+            goto error;
+            }
+      
+      converterQuality = kAudioConverterQuality_Max;
+      status = AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_RenderQuality, kAudioUnitScope_Global, 0, &converterQuality, sizeof(converterQuality));
+      if (status) {
+            qDebug("Failed to set render quality");
+            goto error;
+            }
+      
+      status = changeDesiredStreamFormat(audioUnit, ca->sampleRate());
+      if (status) {
+            qDebug("Failed to set stream format");
+            goto error;
+            }
+    
+      renderCallbackData.inputProc = CoreAudioIOProc;
+      renderCallbackData.inputProcRefCon = ca;
+
+      status = AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_SetRenderCallback, kAudioUnitScope_Output, 0, &renderCallbackData, sizeof(renderCallbackData));
+      if (status) {
+            qDebug("Failed to set render callback");
+            goto error;
+            }
+    
+      AudioUnitInitialize(*audioUnit);
+
+      return true;
+   error:
+      if (*audioUnit) {
+            AudioComponentInstanceDispose(*audioUnit);
+            *audioUnit = NULL;
+            }
+      
+      *audioDevice = kAudioDeviceUnknown;
+      return false;
+      }
+
+static OSStatus CoreAudioIOProc( void* inRefCon,
+                               AudioUnitRenderActionFlags* ioActionFlags,
+                               const AudioTimeStamp* inTimeStamp,
+                               UInt32 inBusNumber,
+                               UInt32 inNumberFrames,
+                               AudioBufferList* ioData)
+      {
+      seq->setInitialMillisecondTimestampWithLatency();
+
+      long bytesPerFrame = sizeof(float) * ioData->mBuffers[0].mNumberChannels;
+
+      float* buffer = static_cast<float*>(ioData->mBuffers[0].mData);
+      UInt64 frames = ioData->mBuffers[0].mDataByteSize / bytesPerFrame;
+
+      if (seq->driver())
+            seq->process((unsigned)frames, buffer);
```

Comment:
please avoid C-style casts. This here may no be needed at all, `frames` being unsigned already?